onnxruntime/core/optimizer/transformer_memcpy.cc

```diff
@@ -184,12 +184,15 @@ void TransformerMemcpyImpl::ProcessDefs(onnxruntime::Node& node, const KernelReg
             initializers_consumed[arg.Name()] = initializer_tensor_proto;
           }
 
-          // implicit inputs have no location info in the kernel def, so treat them as provider inputs.
+          // implicit inputs have no location info in the kernel def, so do nothing to them here, leaving the control
+          //   flow op (Loop, Scan, If) to do the necessary copy if the input crosses different provider.
           // PlannerImpl::ComputeUseCounts has matching logic so the allocation plan does the same thing
-          if (!is_implicit_input && kci && kci->kernel_def->IsInputOnCpu(index)) {
-            non_provider_input_defs_.insert(&arg);
-          } else {
-            provider_input_defs_.insert(&arg);
+          if (!is_implicit_input) {
```

Comment:
Can you clarify how this work? Essentially the implicit input is treated as a graph input to the subgraph. With the main graph inputs we do this:

```
  for (auto arg : graph_.GetInputs())
    // For inputs we need to create a copy node only when the input is connected to both provider
    // and non-provider nodes. Otherwise utils::CopyInputsAcrossDevices() will do the job.
    if (provider_input_defs_.count(arg) && non_provider_input_defs_.count(arg)) {
      AddCopyNode(const_cast<onnxruntime::NodeArg*>(arg), true);
      modified = true;
    }
```

Why is it we can do nothing with an implicit input?

Reply:
`graph_.GetInputs()` does not count implicit inputs, and this code snippet does nothing to implicit inputs.

implicit inputs are dealt inside the control flow OP's implementation. `If`, for example, use feeds fetch manager to copy tensors if necessary.

Reply:
If a single input is used on two different devices we have to manually create a copy node so that the input is available on both devices. It will have a different name on each (original name on one, name of the output from the copy node on the other). For the main graph, GetInputs is fine. For a subgraph I would have expected we need to process implicit inputs as well and potentially insert copy nodes into the subgraph.

The logic in the control flow op implementation is the same that is used for the main graph. It will copy an input to a single device if needed. It does not insert the copy node so a single input can be used on multiple devices.

Reply:
> If a single input is used on two different devices we have to manually create a copy node so that the input is available on both devices. It will have a different name on each (original name on one, name of the output from the copy node on the other). For the main graph, GetInputs is fine.

This is no problem.

> For a subgraph I would have expected we need to process implicit inputs as well and potentially insert copy nodes into the subgraph.

Yes we do. But the way we are doing is different explicit inputs. Specifically, no memcpy nodes will be created for implicit inputs.

> The logic in the control flow op implementation is the same that is used for the main graph.

This can be correct or incorrect depending on how you explain 'the same'. It's different code path.

> It will copy an input to a single device if needed.

This is exactly where the problem happens. The feeds fetch manager rely on `utils::CalculateStaticCopyInfoForFeeds()` to determine the source location of a tensor, which is previously set in `ComputeUseCounts()`. The purpose of this PR is to fix the wrong location set.



Why the change in allocation_planner.cc helps to resolve the issue is explained above in the "About the fix" section: inserting memcpy nodes does not work for implicit input , so setting locations for implicit inputs may cause a wrong result.

Reply:
If I understand correctly this fixes more use cases but still leaves unresolved the scenario where an implicit input is used on two different locations within a single subgraph. The copy across devices is still only putting the implicit in one location for a given subgraph right? I guess we can punt on that fix until it comes up. May be good to note it in the comments that we don't handle this situation. 

Also can we just copy the two new lines of comments from transformer_memcpy.cc to allocation_planner.cc? I think that describes the situation much better than 'setting location for implicit inputs may cause inconsistency in nested subgraph'.

Reply:
you are right. there is potential unresolved scenarios. to completely resolve this issue, we need to apply the transformer/partitioner to the whole graph in one time, instead of doing this for each subgraph recursively. this need a fundamental change.